src/js/plugins/tts/PageChunkIterator.js

```diff
@@ -0,0 +1,153 @@
+import PageChunk from './PageChunk.js';
+
+/**
+ * Class that iterates over the page chunks of a book; caching/buffering
+ * as much as possible to try to ensure a smooth experience.
+ */
+export default class PageChunkIterator {
+    /**
+     * @param {number} pageCount total number of pages
+     * @param {number} start page to start on
+     * @param {PageChunkIteratorOptions} opts
+     */
+    constructor(pageCount, start, opts) {
+        this.pageCount = pageCount;
+        this.opts = Object.assign({}, DEFAULT_OPTS, opts);
+        /** Position in the chunk sequence */
+        this._cursor = { page: start, chunk: 0 };
+        /** @type {Object<number, PageChunk[]>} leaf index -> chunks*/
+        this._bufferedPages = {};
+        /** @type {Object<number, PromiseLike<PageChunk[]>} leaf index -> chunks*/
+        this._bufferingPages = {};
+        /**
+         * @type {Promise} promise that manages cursor modifications so that they
+         * happen in order triggered as opposed to order the server responds
+         **/
+        this._cursorLock = Promise.resolve();
+    }
+
+    /**
+     * Get the next chunk
+     * @return {PromiseLike<"__PageChunkIterator.AT_END__" | PageChunk>}
+     */
+    next() {
+        return this._cursorLock = this._cursorLock
+        .then(() => this._nextUncontrolled());
+    }
+
+    /**
+     * Sends the cursor back 1
+     * @return {Promise}
+     **/
+    decrement() {
+        return this._cursorLock = this._cursorLock
+        .then(() => this._decrementUncontrolled());
+    }
+
+    /**
+     * Gets without ensuring synchronization
+     * @return {PromiseLike<"__PageChunkIterator.AT_END__" | PageChunk>}
+     */
+    _nextUncontrolled() {
+        if (this._cursor.page == this.pageCount) {
+            return Promise.resolve(PageChunkIterator.AT_END);
+        }
+
+        this._recenterBuffer(this._cursor.page);
+
+        return this._fetchPageChunks(this._cursor.page)
+        .then(chunks => {
+            if (this._cursor.chunk == chunks.length) {
+                this._cursor.page++;
+                this._cursor.chunk = 0;
+                return this._nextUncontrolled();
+            }
+            return chunks[this._cursor.chunk++];
+        });
+    }
+
+    /**
+     * Decrements without ensuring synchronization
+     * @return {Promise}
+     */
+    _decrementUncontrolled() {
+        if (this._cursor.page == 0 && this._cursor.chunk == 0) {
+            return this._fetchPageChunks(this._cursor.page);
+        } else if (this._cursor.page > 0 && this._cursor.chunk == 0) {
+            this._cursor.page--;
```

Comment:
suggestion: space between `--`

Reply:
taking a look at this whole block, there may be opportunity to refactor this succinctly as
`return this._fetchPageChunks(this._cursor.page);` is used, with decrement & callback as optional behavior.

Reply:
I'm not sure I follow the refactor; could you expand?

Reply:
Fixed! I did something more promise-y though :P Also opting out of the `--` change; it's inconsistent with the code style of the repo.

Reply:
look so good!